File: Leetcode/98 - Validate Binary Search Tree/solution.ts
function isValidBST(root: TreeNode | null): boolean {
    function validate(node: TreeNode | null, minNode: TreeNode | null, maxNode: TreeNode | null): boolean {
        if (!node) return true;

        if ((minNode && (node.val <= minNode.val)) || (maxNode && (node.val >= maxNode.val))) {
            return false;
        }

        return validate(node.left, minNode, node) && validate(node.right, node, maxNode);
    }

    return validate(root, null, null);
};

File: Leetcode/98 - Validate Binary Search Tree/solution.py
def isValidBST(root: Optional[TreeNode]) -> bool:
    def validate(node: Optional[TreeNode], min_node: Optional[TreeNode], max_node: Optional[TreeNode]) -> bool:
        if not node:
            return True

        if (min_node and node.val <= min_node.val) or (max_node and node.val >= max_node.val):
            return False

        return validate(node.left, min_node, node) and validate(node.right, node, max_node)

    return validate(root, None, None)